client/src/js-sdk/sdk.js

```diff
@@ -1,49 +1,271 @@
 import db, { FIREBASE_FUNCTION_BASE_URL } from './fire.js';
+import firebase from 'firebase'
 import axios from 'axios';
+import uuidv4 from 'uuid/v4';
 
 // Listeners
-export function listenToZone(player, zone, callback) {
-  db.doc(`Games/game1/Players/${player}/Zones/${zone}`)
+export function listenToZone(gameId, player, zone, callback) {
+  db.doc(`Games/${gameId}/Players/${player}/Zones/${zone}`)
     .onSnapshot((doc) => {
       callback(doc.data())
     });
 }
 
-export function listenToPlayer(player, callback) {
-  db.doc(`Games/game1/Players/${player}`)
+export function listenToPlayer(gameId, player, callback) {
+  db.doc(`Games/${gameId}/Players/${player}`)
     .onSnapshot((doc) => {
       callback(doc.data())
     });
 }
 
 // Game functions
-export function tap(card) {
+// Card functions
+export function moveCardToZone(gameId, card, targetZone) {
+  if (card["state.zone"] === targetZone) {
+    return;
+  }
+
+  remove(gameId, card); // remove from current zone before moving
+  var playerName = card["state.owner"];
+  card["state.zone"] = targetZone
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${targetZone}`)
+    .set({
+      [card.id]: card
+    }, { merge: true })
+}
+
+function setSingleCardPosition(gameId, card, newPosition) {
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  card["state.position"] = newPosition
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`)
+    .update({
+      [card.id]: card
+    })
+}
+
+export function setCardPosition(gameId, card, newPosition) {
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  var currentZoneData;
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`).get().then(function(doc) {
+    currentZoneData = doc.data();
+    // Move all other cards that are now below this one down.
+    var previousPosition = card["state.position"]
+    for (var cardId in currentZoneData) {
+      if (currentZoneData.hasOwnProperty(cardId)) {
+        if (cardId !== card.id) {
+          var cardPosition = currentZoneData[cardId]["state.position"]
+          if (cardPosition > previousPosition && cardPosition <= newPosition) {
+            setSingleCardPosition(gameId, currentZoneData[cardId], cardPosition - 1)
+          }
+          else if (cardPosition >= newPosition && cardPosition < previousPosition) {
+            setSingleCardPosition(gameId, currentZoneData[cardId], cardPosition + 1)
+          }
+        } else {
+          setSingleCardPosition(gameId, currentZoneData[cardId], newPosition)
+        }
+      }
+    }
+  });
+}
+
+/**
+ * Randomly shuffle an array
+ * https://stackoverflow.com/a/2450976/1293256
+ * @param  {Array} array The array to shuffle
+ * @return {String}      The first item in the shuffled array
+ */
+function shuffleArray(array) {
```

Comment:
See [this line](https://github.com/warlocks-of-the-midwest/mystic-get-together/blob/master/firebase/functions/src/game.ts#L120) for an easy shuffle. Also worth reading: https://blog.codinghorror.com/shuffling/

Reply:
My link is busted since I pushed new changes, but the line I used to point at was using [lodash](https://www.npmjs.com/package/lodash). You can add a dependency by running `npm i --save <package name>` (`--save` is redundant with later versions of npm). This will download the package and update `package.json` for you. The README for the package should give instructions on how to import the dependency.